db_master_deploy/deploy.sh

```diff
@@ -0,0 +1,331 @@
+    then
+        for i in "${!foreign_keys[@]}"
+        do
+            echo "CREATE FOREIGN KEY CONSTRAINT $i ON $target_id ..."
+            echo "ALTER TABLE ONLY $target_schema.$target_table ADD CONSTRAINT $i ${foreign_keys[$i]};" | psql -U pgkogis -h localhost -d $target_db &> /dev/null
+        done
+    fi
+}
+
+echo "start ${0##*/} $* (pid: $$)"
+CPUS=`(grep "processor" < /proc/cpuinfo | wc -l) 2>/dev/null` || CPUS=1
+START=$(date +%s%3N)
+
+# check for mandatory arguments 
+if [[ -z "$source_objects" || -z "$target" ]];
+then
+    echo "missing a required parameter (source_db -s and staging -t are required)" >&2
+    exit 1
+fi
+
+# check if we have a valid target
+if [[ ! $targets == *$target* ]]
+then
+    echo "valid deploy targets are: '$targets'" >&2
+    exit 1
+fi
+
+# check db access
+if [[ -z `psql -lqt -h localhost -U pgkogis` ]]; then
+    echo "Unable to connect to database cluster" >&2
+    exit 1
+fi
+
+# check for lockfile
+(set -o noclobber; echo "$locktext" > "$lockfile"  2> /dev/null) || { echo "lockfile found: $lockfile '$(cat $lockfile)'" >&2; exit 1; }
+
+trap 'rm -f "$lockfile"; echo "trap: script aborted" >&2; exit $?' INT TERM EXIT
+
+attached_slaves=$(psql -qAt -h localhost -d postgres -U pgkogis -c "select count(1) FROM pg_stat_replication where state='streaming';")
+
+# loop through source_object values
+for source_object in "${array_source[@]}"; do
+    array=(${source_object//./ })
+    # check source objects
+    if [ "${#array[@]}" -ne "3" -a "${#array[@]}" -ne "1" ]; then
+        echo "table data sources have to be formatted like this: db.schema.table, database sources like this: db" >&2
+        exit 1
+    fi
+    # tables go here
+    if [ "${#array[@]}" -eq "3" ]; then
+        echo "processing table $source_object..."
+        source_db=${array[0]}
+        source_schema=${array[1]}
+        source_table=${array[2]}
+        target_db=${source_db%_*}"_"$target # lubis_3d_master -> lubis_3d_$target
+        target_schema="$source_schema"
+        target_table="$source_table"
+        source_id="$source_db.$source_schema.$source_table"
+        target_id="$target_db.$target_schema.$target_table"
+        array_target_table+=($target_id)
+        array_source_table+=($source_object)
+        array_target_combined+=($target_id)
+        check_table
+        check_source
+        copy_table
+    fi
+    # databases go here
+    if [ "${#array[@]}" -eq "1" ]; then
+        echo "processing database $source_object..."        
+        source_db=${array[0]}
+        target_db=${source_db%_*}"_"$target
+        target_db_tmp=$target_db""_tmp_deploy
+        array_target_db+=($target_db)
+        array_source_db+=($source_object)
+        array_target_combined+=($target_db)
+        check_database
+        check_source
+        bod_create_archive
+        copy_database        
+    fi    
+done
+
+END=$(date +%s%3N)
+echo "master has been updated in $((END-START)) milliseconds"
```

Comment:
Any way to get better time info than milliseconds? (Or is it really always that fast?)